### PR TITLE
feat: [M3-7114] - Allow user to choose resize migration type when resizing Linode

### PR DESCRIPTION
**Pull request by jaalah-akamai**
## Description 📝
This work will introduce `warm` migrations which combine the benefits of `cold` and `live` migrations to greatly reduce customer downtime when resizing Linodes.

> **Important**
The notification and event copy/messaging for `warm` migrations will be updated with API work, but you test the UI flow, payload, feature flags and everything else. I will have to update the cypress tests later with actual copy (9/27)

## Major Changes 🔄

- [x] Add feature flag `unifiedMigrations`

**Resize Flow:**
- [x] Add an optional `migration_type` POST field for the `linode/instances/.../resize` endpoint that should be `warm` or `cold`. Backend defaults to `cold`, but in the UI we are defaulting to `warm` unless the user chooses otherwise.
- [x] Add a radio button group for selecting `warm` or `cold` migrations in the `LinodeResize` flow.
    - [x] If the Linode is powered off, we disable controls (even though warm is selected by default) and proceed with a `cold` migration.
- [x] Update & add cypress E2E tests for changes

**Linode Landing:**
- [x] As a user the flag icon is ambiguous and doesn't give much meaning, therefore we want to remove the icon altogether and provide and updated status where it makes sense. For resizes, we want to change the wording from `running` to `migrating` with a green dot icon

**Other:**
- [x]  Fixed the animations for `<Notice />` and fixed a bug in our `fadeIn` keyframes. _Pesky colons!_
    - [x]  In addition, the padding for the `<Notice />` didn't seem right or match `<DismissibleBanner />` which seemed to be the intention, so this has been fixed.

## TODO:
- [ ] Waiting on APIv4 dev work with event/notification messaging to update Cypress tests
- [x] Waiting on UX copy for messaging

## Preview 📷
| Before  | After   |
| ------- | ------- |
| Screenshot 2023-09-13 at 9 30 01 PM | Screenshot 2023-09-13 at 9 29 52 PM |

|Event Notification|
| ------- | 
|Screenshot 2023-09-15 at 11 56 01 AM|

| Enabled  | Disabled   |
| ------- | ------- |
| Screenshot 2023-09-28 at 9 29 14 AM | Screenshot 2023-09-28 at 9 27 39 AM |

| Payload  |
| ------- |
|Screenshot 2023-09-13 at 3 49 32 PM|

| Linode Landing  |
| ------- |
|Screenshot 2023-09-27 at 10 34 27 PM Screenshot 2023-09-27 at 10 30 26 PM|

| Feature Flag  |
| ------- |
|Screenshot 2023-09-25 at 7 45 22 PM|

## How to test 🧪
1. Enable `Unified Migrations` feature flag via dev tools
2. Go to Linodes landing page and resize a Linode from the action menu
3. Observe you now have a radio group to choose `warm` or `cold` migrations with `warm` set as the default
4. When pressing `Resize Linode` you should see a confirmation dialog appear before resize takes effect
5. During resize we should see resize events as outlined in screenshot and notification (notice) on Linode detail page.
    1. Note: The copy for a `warm` resize has not been updating yet (awaiting API), but you should see the payload with the new `migration_type: warm` as shown in screenshot.
6. A `cold` resize does NOT have any additional confirmation
7. Confirm the confirmation dialog closes on error + the resize panel scrolls to the error for a warm migration; confirm that the resize panel scrolls to the error for a cold migration
8. Confirm that errors do not persist upon closing/reopening the resize panel (for both the same linode and a different linode)
9. Confirm that if a linode is offline, default is cold migration + the panel is disabled and a tooltip appears. Confirm that even if you close/reopen the resize panel, the default stays as the cold migration
10. Confirm the feature flag works as expected
11. Confirm the 'Migration Types' label in the migrations panel is not highlighted when choosing between the migration types

### CODE REVIEW

**Comment by mjac0bs** [comment]
> In addition, the padding for the <Notice /> didn't seem right or match <DismissibleBanner /> which seemed to be the intention, so this has been fixed.

Thanks for fixing this! 🎉 🥇  I'd recently noticed that a few notices were looking weirdly skinny, seemingly randomly, but hadn't tracked it down or made a ticket.

File: packages/manager/cypress/e2e/core/linodes/resize-linode.spec.ts [hunk 1]
```diff
@@ -16,9 +26,88 @@ describe('resize linode', () => {
...
       cy.wait('@linodeResize');
+
+      // TODO: Unified Migration: [M3-7115] - Replace with copy from API '../notifications.py'
+      cy.contains(
+        'Your Linode will soon be automatically powered off, migrated, and restored to its previous state (booted or powered off).'
+      ).should('be.visible');
+    });
+  });
+
+  it.only('resizes a linode by increasing size: cold migration', () => {
```
Comment (jdamore-linode) [inline packages/manager/cypress/e2e/core/linodes/resize-linode.spec.ts:42]:
```suggestion
  it('resizes a linode by increasing size: cold migration', () => {
```

**Review by jdamore-linode — COMMENTED** [review]
Hoping to circle back soon with a proper review, but just swinging in now to thank you for the Cypress test additions!

File: packages/manager/cypress/e2e/core/linodes/resize-linode.spec.ts (hunk 1 above)
Comment (jaalah-akamai) [reply packages/manager/cypress/e2e/core/linodes/resize-linode.spec.ts:42]: Thanks! Forgot to revert this from testing!

File: packages/api-v4/src/linodes/actions.ts [hunk 2]
```diff
@@ -77,10 +77,13 @@ export const linodeShutdown = (linodeId: number | string) =>
...
  *
  * @param linodeId { number } The id of the Linode to resize.
- * @param type { string } the new size of the Linode
- * @param auto_resize_linode { boolean } do you want to resize your disks after
+ * @param data { object }
+ * @param data.type { string } the new size of the Linode
+ * @param data.allow_auto_disk_resize { boolean } do you want to resize your disks after
  * the Linode is resized? NOTE: Unless the user has 1 ext disk or 1 ext disk and
- * 1 swap disk, this flag does nothing, regardless of whether it's true or false
+ * 1 swap disk, this flag 4does nothing, regardless of whether it's true or false
```
Comment (coliu-akamai) [inline packages/api-v4/src/linodes/actions.ts:84]:
```suggestion
 * 1 swap disk, this flag does nothing, regardless of whether it's true or false
```

**Review by coliu-akamai — COMMENTED** [review]
✅ Observe you now have a radio group to choose warm or cold migrations 
✅ Warm is set as default
✅  When pressing Resize Linode you should see a confirmation dialog appear before resize takes effect 
✅ During resize we should see resize events as outlined in screenshot and notification (notice) on Linode detail page. Note: The copy for a warm resize has not been updating yet (awaiting API), but you should see the payload with the new migration_type: warm as shown in screenshot. 
✅ A cold resize does NOT have any additional confirmation (assuming this means no confirmation dialog)

Something interesting to note: when I clicked 'Continue' in the confirmation popup, if an error appears, the popup stays -- not sure if this is the behavior you want or not:
image

**Review by abailly-akamai — CHANGES_REQUESTED** [review]
Nice! looks great overall.

A couple comments on the UI:

This tooltip should be wider. The copy is a bit lengthy and It's pretty hard to read currently.
Screenshot 2023-09-27 at 17 30 32

The logic for disabling the field if a linode is powered down is implemented, however there's no explanation as to why. Yes, it's logical we can't perform a warm migration on a powered down Linode, but I don't think we should assume a user will immediately figure out (it's a new feature). I think the UX is lacking here and I'd rather see the "Cold" selection enabled with no option to switch to the warm option and some helper text as to why.

**Comment by jaalah-akamai** [comment]
@abailly-akamai:
- I finished updating our `TooltipIcon` to be more flexible and allow a `width` to be passed in for variable widths, and then I realized the latest mock excludes a tooltip altogether. Anyways I kept it in because I think it could be useful and added a storybook example. 
- I agree with your second concern and I sent it over to UX. We either need some explanation or we just set it to `cold` in the disabled state. I'm awaiting their reply! Hopefully it doesn't hold up this PR because it will need to be in next release. 🤞 

Screenshot 2023-09-27 at 9 20 21 PM

@coliu-akamai The dialog should close on error and scroll to the top still, good catch 👍

**Comment by jaalah-akamai** [comment]
@abailly-akamai I also want to point out some additional changes for the Linodes Landing page (updated description):

- [47cc795](https://github.com/linode/manager/pull/9677/commits/47cc795ffac9f0db275504267e749e9d7a3ceac6)
- [72407cf](https://github.com/linode/manager/pull/9677/commits/72407cfb42fa5bf5b1ae66a21ea22bcbf0c991b8)

**Comment by coliu-akamai** [comment]
Will be checking these off as I go through them
- [ ] Observe you now have a radio group to choose warm or cold migrations
- [ ] Warm is set as default
- [ ] When pressing Resize Linode you should see a confirmation dialog appear before resize takes effect
- [ ] During resize we should see resize events as outlined in screenshot and notification (notice) on Linode detail page. Note: The copy for a warm resize has not been updating yet (awaiting API), but you should see the payload with the new migration_type: warm as shown in screenshot.
- [ ] A cold resize does NOT have any additional confirmation (assuming this means no confirmation dialog)
- [ ] dialog closes on error
- [ ] if linode is offline, default is cold migration + panel disabled

**Comment by coliu-akamai** [comment]
Are errors supposed to be cleared if we close the panel and then reopen? The following error is persistent (which honestly I could see making sense for at least this specific error but maybe not other errors -- just wanted to double check):
image

**Comment by abailly-akamai** [comment]
@jaalah thx for the changes - looks much better 👍 

I also noticed there's an issue with persisting state in the resize modal which I think is unrelated to this PR, but which this PR makes more evident. See video below, look at the persisting error banner and new modal across various resizes attempts

https://github.com/linode/manager/assets/130582365/b7ba9001-db01-466d-b14a-810a76edb1f9

File: packages/api-v4/.changeset/pr-9677-added-1695693363239.md [hunk 3]
```diff
@@ -0,0 +1,5 @@
+---
+"@linode/api-v4": Added
+---
+
+New event type `linode_resize_warm_create` and payload option `migration_type` ([#9677](https://github.com/linode/manager/pull/9677))
```
Comment (dwiley-akamai) [inline packages/api-v4/.changeset/pr-9677-added-1695693363239.md:5]:
```suggestion
 New payload option `migration_type` in `ResizeLinodePayload` and new event type `linode_resize_warm_create`([#9677](https://github.com/linode/manager/pull/9677))
```

File: packages/manager/src/features/Linodes/LinodesDetail/LinodeResize/LinodeResize.tsx [hunk 4]
```diff
@@ -189,6 +251,74 @@ export const LinodeResize = (props: Props) => {
...
+            }}
+          >
+            <Typography variant="h2">Choose Your Resize Type</Typography>
+            <Box
+              sx={{
+                marginTop: 0,
+              }}
+              component="p"
+            >
+              During a <strong>warm resize</strong>, your Compute Instance will
```
Comment (dwiley-akamai) [inline packages/manager/src/features/Linodes/LinodesDetail/LinodeResize/LinodeResize.tsx:274]: We're inconsistent in the casing for "warm resize" in the UI copy -- should it be standardized as either "Warm resize" or "warm resize" across the board?

**Review by dwiley-akamai — COMMENTED** [review]
- Feature flag functionality ✅
- Radio button group for selecting warm or cold migration in the Linode Resize modal ✅
  - Radio button group disabled if Linode is powered off ✅
  - Confirmation modal displayed if “Warm resize” chosen ✅
- Events/Notifications appear as expected ✅

When you click one of the radio buttons, the label gets highlighted:

https://github.com/linode/manager/assets/114682940/eb88001b-5718-4746-ac04-45c30e50a6ae

I'm also observing linodes not having their statuses updated properly after resizes in the Linodes landing table (e.g., resize one that was running --> the process completes and the linode shows as "Offline" in the table, although it's actually running; it doesn't update properly until a refresh), but that seems to be an existing issue?

**Review by coliu-akamai — COMMENTED** [review]
Confirmed that the issue with the error persisting is gone!
Holding off on approval while we're investigating this weird issue some more:

image

File: packages/api-v4/.changeset/pr-9677-added-1695693363239.md (hunk 3 above)
Comment (abailly-akamai) [reply packages/api-v4/.changeset/pr-9677-added-1695693363239.md]: done

File: packages/manager/src/features/Linodes/LinodesDetail/LinodeResize/LinodeResize.tsx (hunk 4 above)
Comment (coliu-akamai) [reply packages/manager/src/features/Linodes/LinodesDetail/LinodeResize/LinodeResize.tsx]: ux confirmed -- should be lowercase across the board (except for the radio button labels - those will say 'Warm resize' or 'Cold resize'), so made those changes!

**Review by abailly-akamai — APPROVED** [review]
Leaving my approval here tho it shouldn't count as one of the co-author.

Tested all scenari i could think of with a special focus on the dialog state management.

**Review by coliu-akamai — APPROVED** [review]
(Same boat with Alban, this approval probably shouldn't count)

✅ Observe you now have a radio group to choose `warm` or `cold` migrations with `warm` set as the default
✅ When pressing `Resize Linode` you should see a confirmation dialog appear before resize takes effect
✅ During resize we should see resize events as outlined in screenshot and notification (notice) on Linode detail page. Note: The copy for a `warm` resize has not been updating yet (awaiting API), but you should see the payload with the new `migration_type: warm` as shown in screenshot.
✅ A `cold` resize does NOT have any additional confirmation
✅  Confirm the confirmation dialog closes on error + scrolls
✅  Confirm that errors do not persist upon closing/reopening the resize panel (for both the same linode and a different linode)
✅  Confirm that if a linode is offline, default is cold migration + panel disabled. Confirm that even if you close/reopen the resize panel, the default stays as the cold migration
✅  Confirm the feature flag works as expected
✅  Confirm the 'Migration Types' label in the migrations panel is not highlighted when choosing between the migration types

Creating a separate ticket for styling of disabled radio buttons in dark theme
image

**Review by bnussman-akamai — APPROVED** [review]
Tested general functionality and everything worked!

I did notice a few things (non-blocking 🚧)
- I don't like that the Migration type is in its own Paper
- The Linode row shows the text "Migrating" instead of "Resizing" during resize
Screenshot 2023-09-28 at 4 14 28 PM


- The general error banner is extra thick
Screenshot 2023-09-28 at 4 10 31 PM

**Review by dwiley-akamai — APPROVED** [review]
Errors don't persist between dialog openings/closings ✅
UI bug of radio group label fixed ✅

Going ahead and approving this but we may need a follow-up ticket/PR for the discrepancies below:

> [x] As a user the flag icon is ambiguous and doesn't give much meaning, therefore we want to remove the icon altogether and provide and updated status where it makes sense. For resizes, we want to change the wording from `running` to `migrating` with a green dot icon

Just to confirm, was this supposed to be "Resizing" --> "Migrating"? In prod we have "Resizing" with a blinking yellow dot. 

Also, should the dot on the Linode Detail page be updated to match what's in the Linodes landing table when the migration is happening? Right now we have "Migrating (xy%)" in the Linodes landing table but the Linode Detail page displays "Resizing":

<details>
<summary>Screenshots</summary>

Screenshot 2023-09-28 at 4 16 41 PM

Screenshot 2023-09-28 at 4 16 51 PM
</details>

**Comment by abailly-akamai** [comment]
@dwiley-akamai & @bnussman-akamai I am thinking these items should be addressed. Taking a look and asking for feedback.

My understanding is that we're indeed calling this a migration now, but I agree it does not make sense to call it that while everything else is calling it a resize. Consistency should be key

**Comment by cpathipa** [comment]
IMO, it should be consistent, Also in linodes details page the font style is all caps ? it should be "Resizing"
image

**Comment by abailly-akamai** [comment]
@mjac0bs @jaalah-akamai 

just to confirm - do we really want to make ALL our notices that fat?

Screenshot 2023-09-29 at 11 34 38

CC @bnussman-akamai https://github.com/linode/manager/pull/9677#pullrequestreview-1649641530

**Comment by abailly-akamai** [comment]
Notice size change has been reverted in this PR and can be revisited later. We're ready to merge as need be.

**Comment by cpathipa** [comment]
@abailly-akamai QQ: During a cold resize, it mentions that the instance will be shut down. In that case, we expected to see a green status icon or an orange one.? 
image

**Comment by abailly-akamai** [comment]
@cpathipa it does for me

Screenshot 2023-09-29 at 13 07 27

**Comment by cpathipa** [comment]
@abailly-akamai It looks like the issue is inconsistent. You can notice the green status at 2:04 in the clip. Looking into that..

https://github.com/linode/manager/assets/119517080/ca83e6fc-d19e-4831-ad13-58a149be62ed

**Comment by bnussman-akamai** [comment]
@cpathipa This may be happening because we don't invalidate Linodes when an event is `started`. I don't know a way off the top of my head to solve this without overfetching

https://github.com/linode/manager/blob/develop/packages/manager/src/queries/linodes/events.ts#L23-L25

File: packages/api-v4/src/linodes/types.ts [hunk 5]
```diff
@@ -394,10 +394,14 @@ export interface LinodeDiskCreationData {
...
 }
 
+export type MigrationTypes = 'warm' | 'cold';
+
 export interface ResizeLinodePayload {
   type: string;
   /** @default true */
   allow_auto_disk_resize?: boolean;
+  /** @default 'warm' */
+  migration_type?: MigrationTypes;
```
Comment (bnussman-akamai) [inline packages/api-v4/src/linodes/types.ts:404]: Which is the actual correct default?

Screenshot 2023-10-02 at 1 15 05 PM

Based on this comment in the PR description, it should be `cold`

Comment (abailly-akamai) [reply packages/api-v4/src/linodes/types.ts:404]: correct, only the ui defaults to warm. will update

**Review by jdamore-linode — APPROVED** [review]
Pushed a quick fix for the one failing resize test! Looking good to me 👍

**Comment by jaalah-akamai** [comment]
Screenshot 2023-10-10 at 9 29 04 AM

- Moved `(Recommended)` outside of tooltip next to label text
- Added `Learn more.` link to documentation for both cold/warm

**Pull request merged**